Ideascale importer

danielSanchezQ opened the pull request:
Use ideascale API to automate creation of json files with the content necessary for the initialization of VSS and voteplan proposal.

This PR supersede https://github.com/input-output-hk/vit-servicing-station/pull/172 (which will be deprecated).

eugene-babichenko commented on src/bin/cli/ideascale/mod.rs:
```diff
@@ -0,0 +1,127 @@
+            .build()?;
+
+        let idescale_data =
+            futures::executor::block_on(runtime.spawn(fetch_all(*fund, api_token.clone())))??;
```
Just this maybe?

```suggestion
            runtime.block_on(fetch_all(*fund, api_token.clone()))??;
```

danielSanchezQ replied:
Well, that is better indeed.

dkijania commented:
I've tested export produced by importer and has a few remarks:

**Fund**

- id: should not be random, it should be strictly defined and equal to fund number
- goal: it should be defined in separate file since it is constant, but may change in the future for example what i have in resources is:

```
[{
   "id":5,
   "goal":"Create, fund and deliver the future of Cardano.",
   "rewards_info":"Create, fund and deliver the future of Cardano.",
   "threshold":450
}]
``` 

**Challenges**

- id: as above it should be non random number strictly increasing starting from 1
- title: please remove 'F5:' (in future it could be F6 F7 and so on)
- lack of proposers_rewards field - need to consult with @MarklarIO if this is something we get from ideascale or in separate file

**Proposals**

- id: as above it should be non random number strictly increasing starting from 1
- proposal_impact_score:  it is always 0. is it something we get from ideascale ?
- missing challenge_id
- missing challenge_type
- proposal_solution is empty

mzabaluev commented on src/bin/cli/ideascale/mod.rs:
```diff
@@ -0,0 +1,138 @@
+    #[structopt(long)]
+    fund_goal: String,
+
+    /// ideascale API token
+    #[structopt(long)]
+    api_token: String,
```
It's somewhat better security to pass the token via an environment variable rather than expose it on the command line. In CI environments this could come from a secret, be screened in logs etc.

danielSanchezQ replied:
Actually yes, let me let it load from both, since it is quite easy to do.

mzabaluev commented on src/bin/cli/ideascale/mod.rs:
```diff
@@ -0,0 +1,138 @@
+
+    /// Path to folder where fund, challenges and proposals json files will be dumped
+    #[structopt(long)]
+    save_folder: PathBuf,
```
Let's use more traditional CLI terminology here: `--output-dir`?

mzabaluev commented on src/bin/cli/ideascale/mod.rs:
```diff
@@ -0,0 +1,138 @@
+use std::str::FromStr;
+
+#[derive(Debug)]
+pub enum VoteType {
```
This could reuse `vote::PayloadTypeDef` from jormungandr-lib.

danielSanchezQ replied:
It is not exposed, iirc.

danielSanchezQ replied:
It is not exposed actually

mzabaluev commented on src/ideascale/fetch.rs:
```diff
@@ -0,0 +1,113 @@
+#[derive(Debug, Deserialize)]
+struct Score {
+    #[serde(alias = "ideaId")]
+    id: i32,
```
Not using `u32` means the idea ID could be negative?

danielSanchezQ replied:
No it can not, it was just most convenient that way since database already work with signed intergers so not to cast them one another.

danielSanchezQ replied:
Actually, those ids are not used finally except for internal linking of data, so I can fairly use unsigned ones.

mzabaluev commented on src/ideascale/fetch.rs:
```diff
@@ -0,0 +1,113 @@
+        .unwrap()
+});
+
+static CLIENT: Lazy<reqwest::Client> = Lazy::new(reqwest::Client::new);
```
I'd rather use an explicit state object to instantiate the client with the API token, and make those fetch functions its methods.
Should not make much difference for the CLI tool, though.

danielSanchezQ replied:
Yes, but it is messy to spawn clients in the async tasks later. It will avoid the api_token spreading though.

mzabaluev replied:
API token retrieval from an environment variable could go under the lazy initialization, but that means you can't easily have the command line option for it (which the tool probably shouldn't, as per prior discussion)

mzabaluev commented on src/ideascale/models/de.rs:
```diff
@@ -0,0 +1,152 @@
+    deserializer: D,
+) -> Result<String, D::Error> {
+    let mut rewards_str = String::deserialize(deserializer)?;
+    rewards_str.retain(|c| !DIRTY_CHARACTERS.contains(&c));
```
`rewards_str.retain(|c| !matches!(c, '*' | '-' | '/'))` would work faster, no need for a hash set.

mzabaluev commented on src/ideascale/models/de.rs:
```diff
@@ -0,0 +1,152 @@
+        return Ok("0".to_string());
+    }
+    sscanf::scanf!(rewards_str.trim_end(), "${} in ada", String)
+        // trim all . or , in between numbers
+        .map(|mut s| {
+            s.retain(|c: char| c.is_numeric());
+            s
+        })
```
Is it possible that some of the separators would be a decimal point? This transformation feels a bit too loose for this important amount.

danielSanchezQ replied:
It may be, but they are cut out from input like `$600,000 in ada`, The format is the worry here. Eventually we should have a proper field that is actually a real value.

mzabaluev replied:
Let's restrict this to a format that is in de-facto use: remove only commas. It would be good to validate the resulting string parses as a `u64` integer, because otherwise vit-servicing-station will fail to load.
In fact, I'd make the field type for such ADA values a newtype over `u64`.

danielSanchezQ commented:
@dkijania , it is ready for you to check again I think, thanks! :)


Fund

id: it wasnt random, I used the one in ideascale, it is fixed now ✅ 
goal: got this from a cli argument ✅ 

Challenges

id: same as funds, fixed too. ✅ 
title: please remove 'F5:' (in future it could be F6 F7 and so on) 
lack of proposers_rewards field - this is not needed anymore ✅ 


Proposals

id: likewise, fixed now ✅ 
proposal_impact_score: it is always 0. is it something we get from ideascale ? It was an error from we were getting them, idescale relate ✅ 
missing challenge_id -> added ✅ 
missing challenge_type -> addded ✅ 
proposal_solution is empty -> bug, fixed now ✅

mzabaluev commented on src/ideascale/models/se.rs:
```diff
@@ -0,0 +1,47 @@
+use serde::Serialize;
+
+#[derive(Debug, Serialize)]
+pub struct Challenge {
```
It may become a burden to maintain two sets of structures.
Is it not possible to define `Serialize` and `Deserialize` over a single `Challenge` struct?

danielSanchezQ replied:
I decided to use 2 because data to be deserialize and to be serialize is not the same. Not just the types but also the content itself. Some attributes are just needed to reconstruct part of the data of other structs, but may not appear in the final `de -> se` representation. Is much simpler this way instead of dealing with the serde `optional` and `skip_de/serialize` dance.

danielSanchezQ commented on src/ideascale/mod.rs:
```diff
@@ -0,0 +1,204 @@
+        .filter(|p| filter_proposal_by_stage_type(&p.stage_type))
+        .collect();
+
+    let mut stages: Vec<_> = fetch::get_stages(api_token.clone()).await?;
+    stages.retain(|stage| filter_stages(stage, &funnels));
```
@mzabaluev Proposed solution for scores retrieval.

mzabaluev commented on src/ideascale/mod.rs:
```diff
@@ -0,0 +1,209 @@
+}
+
+fn filter_stages(stage: &Stage, stage_label: &str, funnel_ids: &HashMap<u32, Funnel>) -> bool {
+    stage.label.to_ascii_lowercase() == stage_label.to_ascii_lowercase()
```
This allocates a string from `stage_label` every time that could be pre-normalized once instead, in the code calling `filter_stages`.

mzabaluev reviewed (approved):
Looks good. Test failures seem to be unrelated.
